### PR TITLE
Results - cluster view

### PR DESCRIPTION
**Pull request by t0maboro**
Todo:
- [x] Preview cluster elements
- [x] Plot cluster instead of centroids when showing elements
- [x] Save clusters to file

If you think that something in layout/logic should be done differently - let me know. 
To discuss: format of the file for saving cluster. Saving to csv should be done easily, but we might consider other formats, or leave a choice for users.

### CODE REVIEW

File: src/widgets/results_widgets/k_means_results.py [hunk 1]
```diff
@@ -132,3 +138,80 @@ def update_plot(self):
         self.centroids_canvas.new_centroids_plot(None, None, x_centroids, y_centroids, self.ox, self.oy,
                                                  min_x - sep_x, max_x + sep_x, min_y - sep_y, max_y + sep_y,
                                                  )
+
+    def update_cluster_plot(self, row=None):
```
Comment (mhawryluk) [inline src/widgets/results_widgets/k_means_results.py]: Maybe rename _row_ to _selected_label_ or _selected_cluster_

File: src/widgets/results_widgets/k_means_results.py [hunk 2]
```diff
@@ -132,3 +138,80 @@ def update_plot(self):
...
+
+    def exit_from_cluster(self):
+        self.centroids_table.setModel(QtTable(self.centroids.round(3)))
+        self.centroids_group_layout.itemAt(0).widget().setParent(None)
+        self.centroids_table_instruction.show()
+        self.centroids_table.doubleClicked.connect(self.show_cluster)
+        self.update_cluster_plot()
+
+    def on_save_button_click(self, elements):
+        path, is_ok = QInputDialog.getText(self, 'input name', 'Enter filename')
```
Comment (mhawryluk) [inline src/widgets/results_widgets/k_means_results.py]:
```suggestion
        path, is_ok = QInputDialog.getText(self, 'Save to file', 'Enter filename')
```

File: src/widgets/results_widgets/k_means_results.py [hunk 3]
```diff
@@ -132,3 +138,80 @@ def update_plot(self):
...
+        if is_ok and path and path.endswith(".csv"):
+            elements.to_csv(path)
+        else:
+            error = QMessageBox()
+            error.setIcon(QMessageBox.Critical)
+            error.setText("An error has occurred while trying to write cluster to file")
+            error.setWindowTitle("Error")
+            error.exec_()
```
Comment (mhawryluk) [inline src/widgets/results_widgets/k_means_results.py]: We shouldn't show an error when the user simply cancelled the window. I assume is_ok is False in that case. Also when the extension is wrong the error message should state that the file type is incorrect. The current message makes sense if there was indeed an exception while trying to create/access the file.

**Review by mhawryluk — CHANGES_REQUESTED** [review]
Looks great, I love the cluster selecting by double clicking on table row, brilliant! 
One thing that doesn't work yet is changing the axes on the chart while in the specific cluster view. It plots all the centroids again like it used to.

**Review by gregori0o — COMMENTED** [review]
That is very good idea to selecting cluster by double-click and show them in Centroids section.

There are some problems with message box and changing axis, which @mhawryluk mentioned.
I suggest add one more functionality. Button 'save result' or something like that, which save data to file with new column. In this column will be number of cluster for the row.

File: src/widgets/results_widgets/k_means_results.py [hunk 4]
```diff
@@ -194,24 +201,40 @@ def show_cluster(self):
...
+                error.setIcon(QMessageBox.Critical)
+                error.setText("Something wrong happened while was writing data to file. Try again")
+                error.setWindowTitle("Saving failed")
+                error.exec_()
+        elif not is_ok:
+            pass
+        elif not path:
+            error = QMessageBox()
+            error.setIcon(QMessageBox.Critical)
+            error.setText("No path were provided")
```
Comment (mhawryluk) [inline src/widgets/results_widgets/k_means_results.py]:
```suggestion
            error.setText("No path was provided")
```

File: src/widgets/results_widgets/k_means_results.py [hunk 5]
```diff
@@ -194,24 +201,40 @@ def show_cluster(self):
...
-        path, is_ok = QInputDialog.getText(self, 'input name', 'Enter filename')
+        path, is_ok = QInputDialog.getText(self, 'Save to file', 'Enter filename')
         if is_ok and path and path.endswith(".csv"):
-            elements.to_csv(path)
+            try:
+                elements.to_csv(path)
+            except:
+                error = QMessageBox()
+                error.setIcon(QMessageBox.Critical)
+                error.setText("Something wrong happened while was writing data to file. Try again")
```
Comment (mhawryluk) [inline src/widgets/results_widgets/k_means_results.py]:
```suggestion
                error.setText("Something wrong happened while writing data to file. Try again")
```

File: src/widgets/results_widgets/k_means_results.py [hunk 6]
```diff
@@ -175,10 +183,9 @@ def update_cluster_plot(self, row=None):
                                                      )
 
     def show_cluster(self):
-        cluster_num = 0
         for idx in self.centroids_table.selectionModel().selectedIndexes():
-            cluster_num = idx.row()
-        rows = [i for i in range(len(self.labels)) if self.labels[i] == cluster_num]
+            self.selected_cluster = idx.row()
```
Comment (mhawryluk) [inline src/widgets/results_widgets/k_means_results.py]: it seems as though this doesn't have to be a loop

**Review by mhawryluk — APPROVED** [review]
Great, just a few tiny suggestions :)

**Review by gregori0o — APPROVED** [review]
❤️ 💪

**Pull request merged**